### PR TITLE
Use exchange package's Exchange implementation in JSON tests

### PR DESCRIPTION
**Pull request by guscarreon**
The current version of prebid server uses a `mockBidExchange` object in order to run the `TestJsonSampleRequests` test found in `endpoints/openrtb2/auction_test.go`. This is a very comprehensive test that reads input and expected data from the JSON files found inside the `endpoints/openrtb2/sample-requests` directory. This pull request removes `mockBidExchange` and uses the `exchange/` directory's implementation of an `exchange` instead. To do so, this PR:
  1. Exports the `AdaptedBidder` interface defined inside `exchange/bidder.go` in order to build an adapter map inside `endpoints/openrtb2/auction_test.go` whose elements are mock bidder adapters pointing to mock bidder servers.
  2. Makes the appropriate changes inside `endpoints/openrtb2/auction_test.go` to drop `mockBidExchange` and use an `exchange` implementation.
  3. Incorporates file `endpoints/openrtb2/test_utils.go` in order to break the enormous `endpoints/openrtb2/auction_test.go` file into two. `endpoints/openrtb2/test_utils.go` was written to carry auxiliary types, functions and interface implementations useful for the testing framework.
  4. Tweaks JSON files where needed. For instance, the following SON test files were modified to expect warnings instead:
  ```
endpoints/openrtb2/sample-requests/currency-conversion/custom-rates/errors/conversion-disabled.json
endpoints/openrtb2/sample-requests/currency-conversion/custom-rates/errors/custom-rates-currency-missing-usepbs-default.json
endpoints/openrtb2/sample-requests/currency-conversion/custom-rates/errors/custom-rates-currency-missing-usepbs-false.json
endpoints/openrtb2/sample-requests/currency-conversion/custom-rates/errors/custom-rates-empty-usepbs-false.json
endpoints/openrtb2/sample-requests/currency-conversion/server-rates/errors/no-conversion-found.json
  ```
  5. Defines `BenchmarkAllExtRequest` inside `endpoints/openrtb2/auction_benchmark_test.go` and corrects a bug in `BenchmarkOpenrtbEndpoint` also found in said file.
  
Please let me know your takes.

### CODE REVIEW

**Comment by VeronikaSolovei9** [comment]
Overall looks good to me, simpler and clearer. 
I have some questions to make sure I understand it all right.

1) With new test exchange there are will be 3 predefined bidders for all exchange tests: appnexus, openx and rubicon, right? 
2) Benchmark tests: are they the part of existing tests or they should be run separately to check memory/cpu usage is still ok after some new changes? 
3) Why existing tests started to throw a warning?

**Comment by guscarreon** [comment]
> 1) With new test exchange there are will be 3 predefined bidders for all exchange tests: appnexus, openx and rubicon, right?

Right I didn't want to create a new bidder simply for testing purposes so these tests simply mock these bidders' bid servers. Do you agree with this approach?

 > 2) Benchmark tests: are they the part of existing tests or should they be run separately to check memory/cpu usage is still ok after some new changes?

They run separately with the `-bench` flag of the `go test` command. 

 > 3) Why existing tests started to throw a warning?

Any warning in particular? The `HoldAuction(ctx context.Context, r AuctionRequest, debugLog *DebugLog)` implementation found in the `exchange` package builds a `bidResponse.Ext` with the contract found in the  `openrtb_ext` package:
```
178 func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *DebugLog) (*openrtb2.BidResponse, error) {
179     var err error
180     requestExt, err := extractBidRequestExt(r.BidRequest)
181 *-- 51 lines: if err != nil {--------------------------------------------------------------------------------------------
232
233     var auc *auction
234     var cacheErrs []error
235     var bidResponseExt *openrtb_ext.ExtBidResponse
236     if anyBidsReturned {
237 *-- 87 lines: var bidCategory map[string]string--------------------------------------------------------------------------
324     // Build the response
325     return e.buildBidResponse(ctx, liveAdapters, adapterBids, r.BidRequest, adapterExtra, auc, bidResponseExt, cacheInstructions.returnCreative, r.ImpExtInfoMap, errs)
326 }
exchange/exchange.go
```
Does this answer your question?

File: endpoints/openrtb2/auction_test.go
Comment (VeronikaSolovei9) [inline endpoints/openrtb2/auction_test.go:127]: Do you think these closing functions can go inside `buildTestEndpoint` func like it was before `defer appNexusServer.Close()`? You will not need to return all servers you need to close, there are no other usages of these variables.

**Comment by VeronikaSolovei9** [comment]
Overall LGTM. Would be good to have another pair of eyes on it to make sure I didn't miss anything

File: endpoints/openrtb2/auction_test.go
Comment (guscarreon) [reply endpoints/openrtb2/auction_test.go:127]: From [this article](https://www.geeksforgeeks.org/defer-keyword-in-golang/): "defer statements delay the execution of the function or method or an anonymous method until the nearby function returns". Therefore, I assume a `defer` statement inside `buildTestEndpoint(test testCase, paramValidator openrtb_ext.BidderParamValidator)` prevents a given server from listening as soon as we exit said function. But I agree with you that the `Close()` statements were misplaced because an error thrown by `buildTestEndpoint(test testCase, paramValidator openrtb_ext.BidderParamValidator)` would prevent us from closing them altogether. I corrected accordingly. Please let me know if you agree with the new approach.

Comment (guscarreon) [reply endpoints/openrtb2/auction_test.go:127]: > Edit: and for auction_benchmark_test.go too.
I agree. Given that from `BenchmarkValidWholeExemplary` we exit with `b.Fatal(err.Error())` in case `buildTestEndpoint` returns an error, the `defer` keyword fits our needs just fine. Corrected.

Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction_test.go:127]: Right, defer will be executed at the closing bracket of the function, but you need to have those servers(not variables) available after buildTestEndpoint. I see, you are right, thank you for checking this!

Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction_test.go:127]: Will approve it if there will be no changes required from other reviewers.

File: endpoints/openrtb2/auction_test.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction_test.go:190]: For my own understanding, did you change this variable name here from `k` to `bidderName` just for clarity? Or was something structurally changed that required a name change.

File: endpoints/openrtb2/auction_benchmark_test.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction_benchmark_test.go:67]: Does this change just initialize this variable?

**Review by AlexBVolcy — COMMENTED** [review]
This is looking good, just had a couple questions about minor things in the code to improve my understanding!

File: endpoints/openrtb2/auction_test.go
Comment (guscarreon) [reply endpoints/openrtb2/auction_test.go:190]: Yes. I was trying to make it more readable. `k` seemed to not be as approachable as `bidderName` to me.

File: endpoints/openrtb2/auction_benchmark_test.go
Comment (guscarreon) [reply endpoints/openrtb2/auction_benchmark_test.go:67]: Exactly. Turns out the current version of this file in prod panics whenever we want to run this benchmark because the map is not explicitly initialized:
```
╭─gcarreongutierrez ~/go/src/github.com/prebid/prebid-server/endpoints/openrtb2  ‹master›
╰─➤  go test -run=XXX -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-server/endpoints/openrtb2
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkOpenrtbEndpoint-12     panic: assignment to entry in nil map

goroutine 50 [running]:
github.com/prebid/prebid-server/endpoints/openrtb2.BenchmarkOpenrtbEndpoint(0xc0000b6000)
        /Users/gcarreongutierrez/go/src/github.com/prebid/prebid-server/endpoints/openrtb2/auction_benchmark_test.go:65 +0x185
testing.(*B).runN(0xc0000b6000, 0x1)
        /usr/local/opt/go/libexec/src/testing/benchmark.go:193 +0x102
testing.(*B).run1.func1()
        /usr/local/opt/go/libexec/src/testing/benchmark.go:233 +0x59
created by testing.(*B).run1
        /usr/local/opt/go/libexec/src/testing/benchmark.go:226 +0x9c
exit status 2
FAIL    github.com/prebid/prebid-server/endpoints/openrtb2      0.441s
```
Explicitly initializing corrects this issue

File: endpoints/openrtb2/auction_test.go
Comment (AlexBVolcy) [reply endpoints/openrtb2/auction_test.go:190]: I think the name change to `bidderName` is great!

File: endpoints/openrtb2/auction_test.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction_test.go:102]: In the PR description, you mention the exporting of the `AdaptedBidder` interface so that a mock bidder adapter to mock bidder server map can be built in this file. I'm having trouble finding this map, could you point it out to me?

Comment (guscarreon) [reply endpoints/openrtb2/auction_test.go:102]: Originally `buildTestEndpoint(test testCase, paramValidator openrtb_ext.BidderParamValidator)` was written into `endpoints/openrtb2/auction_test.go` but in an effort to refactor said file, this function was moved to `endpoints/openrtb2/test_utils.go`. Exported `AdaptedBidder` is instantiated in line 1084 of `endpoints/openrtb2/test_utils.go` shown below:
```
1078 func buildTestEndpoint(test testCase, paramValidator openrtb_ext.BidderParamValidator) (httprouter.Handle, *httptest.Server, *httptest.Server, *httptest.Server, *httptest.Server, error) {
1079     bidderInfos := getBidderInfos(test.Config.getAdaptersConfigMap(), openrtb_ext.CoreBidderNames())
1080     bidderMap := exchange.GetActiveBidders(bidderInfos)
1081     disabledBidders := exchange.GetDisabledBiddersErrorMessages(bidderInfos)
1082
1083     // Adapter map with mock adapters needed to run JSON test cases
1084     adapterMap := make(map[openrtb_ext.BidderName]exchange.AdaptedBidder, 0)
1085
1086     // AppNexus mock bid server and adapter
1087     appNexusBidder := mockBidderHandler{bidInfo: test.Config.MockBidder, bidderName: "appnexus"}
endpoints/openrtb2/test_utils.go
```

**Comment by AlexBVolcy** [comment]
Just want to point out that a test is failing in `TestGoodAmpRequests` for what looks like a seg fault

**Comment by guscarreon** [comment]
> Just want to point out that a test is failing in `TestGoodAmpRequests` for what looks like a seg fault

It's been corrected Alex

**Review by AlexBVolcy — APPROVED** [review]

**Comment by VeronikaSolovei9** [comment]
Gus, what you implemented for amp endpoint testing is impressive! 
I want to confirm couple of things. 
For `TestGoodAmpRequests` you specify 7 test cases. 6 of them have `mockBidRequest.imp.ext.prebid.storedauctionresponse` which means that some of exchange logic will not be executed: `e.getAllBids`, exchange.go, L274 for majority of test cases. Instead `buildStoredAuctionResponse` L265 will be executed. Is this the intention? Or should only `site-stored-response.json` test execute `buildStoredAuctionResponse`? 
To be clear, I'm ok with this, just want to make sure this is how it should be. The only test that executes `getAllBids` is `gdpr.json`. This file still have `mockAmpStoredResponse` despite it's not needed for this test (it works without it, I tried :)).

File: endpoints/openrtb2/amp_auction_test.go
Comment (bsardo) [inline endpoints/openrtb2/amp_auction_test.go:29]: This is looking good. Nice work!
My only suggestions regarding this is to have every JSON file begin with the most simplistic valid base request that would go through the whole flow without using any stored responses. Then slightly tweak each test to verify the specific part that is under test. This would establish symmetry among the tests reducing confusion.

Some of the JSON files contain additional unrequired fields like `targeting`, `pmp.deals`, etc and it isn't clear why those fields are part of that test. If we actually want to make sure all of those unrequired fields don't break the AMP endpoint but they don't require their own dedicated JSON file tests (because it would result in an explosion in the number of JSON files), I suggest creating one JSON test with a fully loaded request, but perhaps even that isn't necessary.

As for establishing symmetry and improving intent, it seems like:
-  only `imp-with-stored-resp.json` should define `mockAmpStoredResponse`.
- `mock`gdpr.json` and `gdpr-no-consentstring.json` should be the same except for the consent string.
- `aliases.json` and `aliased-buyeruids.json` should be the same except for `user.ext.prebid.buyeruids`
- `user.json` and `aliased-buyeruids.json` should be the same except for `ext.prebid.aliases`

**Review by bsardo — COMMENTED** [review]
This is looking good. I have just a few comments.

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:123]: Done

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:171]: Corrected

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:93]: Done

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:1083]: Updated

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:1201]: For the purpose of moving forward with this PR's core functionality I wrote this sloppy version of the `FetchResponses` implementation and then forgot to polish it at the end. Let me know if you agree with the final version.

File: endpoints/openrtb2/amp_auction_test.go
Comment (guscarreon) [reply endpoints/openrtb2/amp_auction_test.go:29]: > `aliases.json` and `aliased-buyeruids.json` should be the same except for `user.ext.prebid.buyeruids`
 
 Should we also swap the `"unknown"` to `"appnexus"` in the `"ext"` too? Or is that part of the test? Is the test trying to prove that PBS core understands that `"unknown"` found in the `"buyeruids"` field refers to `"appnexus"`?
```
19         "ext": {
20           "prebid": {
21               "storedauctionresponse": {
22                   "id": "6d718149"
23               }
24           },
25 -         "appnexus": {
   +         "unknown": {
26             "placementId": 12883451
27           }
28         }
29       }
30     ],
31     "user": {
32       "ext": {
33         "prebid": {
34           "buyeruids": {
35             "unknown": "123"
36           }
37         }
38       }
39     },
40     "ext": {
41       "prebid": {
42         "aliases": {
43           "unknown": "appnexus"
44         }
45       }
46     }
47   },
endpoints/openrtb2/sample-requests/valid-whole/supplementary/aliased-buyeruids.json
```

Comment (guscarreon) [reply endpoints/openrtb2/amp_auction_test.go:29]: > `user.json` and `aliased-buyeruids.json` should be the same except for `ext.prebid.aliases`
 
`user.json` didn't need the user part to pass. When removed, it looked exactly like `endpoints/openrtb2/sample-requests/valid-whole/supplementary/imp-with-stored-resp.json` which is the same as `aliased-buyeruids.json` except for `ext.prebid.aliases`. I removed `user.json`

 > mockgdpr.json and gdpr-no-consentstring.json should be the same except for the consent string.

Agree, reduced.

 > only `imp-with-stored-resp.json` should define `mockAmpStoredResponse`.

Agree, reduced.

`endpoints/openrtb2/sample-requests/valid-whole/supplementary/site-stored-response.json` without the clutter is the same as `endpoints/openrtb2/sample-requests/valid-whole/supplementary/imp-with-stored-resp.json`. The former got removed.

**Comment by guscarreon** [comment]
> Gus, what you implemented for amp endpoint testing is impressive! 

Thanks a lot Veronika 😄

**Comment by guscarreon** [comment]
> I want to confirm couple of things. For `TestGoodAmpRequests` you specify 7 test cases. 6 of them have `mockBidRequest.imp.ext.prebid.storedauctionresponse` which means that some of exchange logic will not be executed: `e.getAllBids`, exchange.go, L274 for majority of test cases. Instead `buildStoredAuctionResponse` L265 will be executed. Is this the intention? Or should only `site-stored-response.json` test execute `buildStoredAuctionResponse`? To be clear, I'm ok with this, just want to make sure this is how it should be. The only test that executes `getAllBids` is `gdpr.json`. This file still have `mockAmpStoredResponse` despite it's not needed for this test (it works without it, I tried :)).
> 
> Other than this it looks good to me!

Agree. All the files were reduced to their minimum number of entries and their description was updated. After the reduction happened, both `endpoints/openrtb2/sample-requests/valid-whole/supplementary/site-stored-response.json` and `endpoints/openrtb2/sample-requests/valid-whole/supplementary/user.json` were left identical to ` "imp-with-stored-resp.json"` and they were removed. Please let me know if you agree with this decision or if you believe another test could complement our set here.

> 6 of them have `mockBidRequest.imp.ext.prebid.storedauctionresponse` which means that some of exchange logic will not be executed: `e.getAllBids`, exchange.go, L274 for majority of test cases. Instead `buildStoredAuctionResponse` L265 will be executed. 

Good point. I removed the `"mockBidders"` or the `mockAmpStoredResponse` entries were it made sense and tried to improve the JSON files `"description"` fields in order to make perfectly clear from where are we pulling the final bid from. Please let me know if you approve the wording.

I hope I didn't forget anything or left any loose ends.

File: endpoints/openrtb2/amp_auction_test.go
Comment (guscarreon) [reply endpoints/openrtb2/amp_auction_test.go:29]: > This is looking good. Nice work!
 
 Thanks Brian 😄

File: endpoints/openrtb2/test_utils.go
Comment (AlexBVolcy) [reply endpoints/openrtb2/test_utils.go:1210]: I'm interested in what PR this would've come from?

**Comment by AlexBVolcy** [comment]
This looks good to me, I'm also interested in what Brian brought up in relation to `mockAmpStoredResponse` and also the new code that somehow came from a merge? After that I'll approve!

File: exchange/bidder.go
Comment (guscarreon) [reply exchange/bidder.go:57]: weird 🤔 . I do `git blame` on that very line and it seems to trace it to its righteous owner:
image

File: endpoints/openrtb2/test_utils.go
Comment (guscarreon) [reply endpoints/openrtb2/test_utils.go:1210]: Discussed offline. This is a new file

File: endpoints/openrtb2/sample-requests/valid-whole/supplementary/site-stored-response.json
Comment (guscarreon) [reply endpoints/openrtb2/sample-requests/valid-whole/supplementary/site-stored-response.json]: I can definitely bring this file back. I discarded it in the last commit because I thought it was too similar to `endpoints/openrtb2/sample-requests/valid-whole/supplementary/imp-with-stored-resp.json`. Do you think they differ enough to bring it back?

Comment (bsardo) [reply endpoints/openrtb2/sample-requests/valid-whole/supplementary/site-stored-response.json]: Discussed offline. There's no need to bring this test back. AMP stored responses are adequately covered by the `imp-with-stored-resp.json` test.

File: endpoints/openrtb2/sample-requests/valid-whole/supplementary/aliased-buyeruids.json
Comment (bsardo) [reply endpoints/openrtb2/sample-requests/valid-whole/supplementary/aliased-buyeruids.json:5]: Discussed offline. We agreed to keep all of these tests but to only use stored responses in `imp-with-stored-resp.json`. The other four tests (two GDPR and two alias) should use the `mockBidders` test feature to ensure we execute as much of the PBS core logic as possible.

File: endpoints/openrtb2/auction_benchmark_test.go
Comment (guscarreon) [reply endpoints/openrtb2/auction_benchmark_test.go]: done

**Review by AlexBVolcy — APPROVED** [review]
LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**
